core/src/vault_registry.rs

```diff
@@ -0,0 +1,433 @@
+        }
+
+        if let Some(switchboard_feed) = switchboard_feed {
+            mint_entry.switchboard_feed = switchboard_feed;
+        }
+
+        if let Some(no_feed_weight) = no_feed_weight {
+            mint_entry.no_feed_weight = PodU128::from(no_feed_weight);
+        }
+
+        self.check_st_mint(st_mint)?;
+
+        Ok(())
+    }
+
+    pub fn register_vault(&mut self, mint: Pubkey, vault_index: u64) -> Result<(), ProgramError> {
+        // Check if (mint, vault_index) is already in the list
+        if self
+            .vault_list
+            .iter()
+            .any(|m| m.st_mint == mint && m.vault_index() == vault_index)
+        {
+            return Ok(());
+        }
+
+        // Check if vault_index is already in use by a different mint
+        if self
+            .vault_list
+            .iter()
+            .any(|m| m.vault_index() == vault_index)
+        {
+            return Err(TipRouterError::VaultIndexAlreadyInUse.into());
+        }
+
+        // Insert at the first empty slot
+        let mint_entry = self
+            .vault_list
+            .iter_mut()
+            .find(|m| m.st_mint == VaultEntry::default().st_mint)
+            .ok_or(TipRouterError::TrackedMintListFull)?;
+
+        *mint_entry = VaultEntry::new(mint, vault_index);
+        Ok(())
+    }
+
+    pub fn get_vault_entries(&self) -> Vec<VaultEntry> {
+        self.vault_list
+            .iter()
+            .filter(|m| !m.is_empty())
+            .copied()
+            .collect()
+    }
+
+    pub fn vault_count(&self) -> u64 {
+        self.vault_list.iter().filter(|m| !m.is_empty()).count() as u64
+    }
+
+    pub fn get_mint_entries(&self) -> Vec<StMintEntry> {
+        self.st_mint_list
+            .iter()
+            .filter(|m| !m.is_empty())
+            .copied()
+            .collect()
+    }
+
+    pub fn get_mint_entry(&self, st_mint: &Pubkey) -> Result<StMintEntry, ProgramError> {
+        let mint_entry = self
+            .st_mint_list
+            .iter()
+            .find(|m| m.st_mint().eq(st_mint))
+            .ok_or(TipRouterError::MintEntryNotFound)?;
+
+        Ok(*mint_entry)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_len() {
+        use std::mem::size_of;
+
+        let expected_total = size_of::<Pubkey>() // ncn
+            + 1 // bump
+            + 127 // reserved
+            + size_of::<StMintEntry>() * MAX_ST_MINTS // st_mint_list
+            + size_of::<VaultEntry>() * MAX_VAULT_OPERATOR_DELEGATIONS; // vault_list
+
+        assert_eq!(size_of::<VaultRegistry>(), expected_total);
+
+        let tracked_mints = VaultRegistry::new(Pubkey::default(), 0);
+        assert_eq!(
+            tracked_mints.vault_list.len(),
+            MAX_VAULT_OPERATOR_DELEGATIONS
+        );
+    }
+
+    #[test]
+    fn test_add_mint() {
+        let mut tracked_mints = VaultRegistry::new(Pubkey::default(), 0);
```

Comment:
tracked mints in a lot of these test names

Reply:
should be addressed